Sources/SwiftFormatCore/RuleMask.swift

```diff
@@ -0,0 +1,119 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import SwiftSyntax
+
+/// This class takes the raw source text and scans through it searching for comment pairs of the
+/// form:
+///
+///   3. |  // swift-format-disable: RuleName
+///   4. |  let a = 123
+///   5. |  // swift-format-enable: RuleName
+///
+/// This class records that `RuleName` is disabled for line 4. The rules themselves reference
+/// RuleMask to see if it is disabled for the line it is currently examining.
+public class RuleMask {
+
+  /// Each rule has a list of ranges for which it is disabled.
+  private var ruleMap: [String: [Range<Int>]] = [:]
+
+  /// Regex to match the enable comments; rule name is in the first capture group.
+  private let enablePattern = #"^\s*//\s*swift-format-enable:\s+(\S+)"#
+
+  /// Regex to match the disable comments; rule name is in the first capture group.
+  private let disablePattern = #"^\s*//\s*swift-format-disable:\s+(\S+)"#
+
+  /// Rule enable regex object.
+  private let enableRegex: NSRegularExpression
+
+  /// Rule disable regex object.
+  private let disableRegex: NSRegularExpression
+
+  /// Used to compute line numbers of syntax nodes.
+  private let sourceLocationConverter: SourceLocationConverter
+
+  /// This takes the head Syntax node of the source and generates a map of the rules specified for
+  /// disable/enable and the line ranges for which they are disabled.
+  public init(syntaxNode: Syntax, sourceLocationConverter: SourceLocationConverter) {
+    enableRegex = try! NSRegularExpression(pattern: enablePattern, options: [])
+    disableRegex = try! NSRegularExpression(pattern: disablePattern, options: [])
+
+    self.sourceLocationConverter = sourceLocationConverter
+    generateDictionary(syntaxNode)
+  }
+
+  /// Calculate the starting line number of a syntax node.
+  private func getLine(_ node: Syntax) -> Int? {
+    let loc = node.startLocation(converter: self.sourceLocationConverter)
+    return loc.line
+  }
+
+  /// Check if a comment matches a disable/enable flag.
+  private func getRule(regex: NSRegularExpression, text: String) -> String? {
+    let nsrange = NSRange(text.startIndex..<text.endIndex, in: text)
+    if let match = regex.firstMatch(in: text, options: [], range: nsrange) {
+      let matchRange = match.range(at: 1)
```

Comment:
`String.range(of:options:)` looks to only take a `StringProtocol` for the regex, not an `NSRegularExpression`, so it would force recompilation of the regex each time it's called.